From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Markus Rudy <mr@edgeless.systems>
Date: Tue, 24 Feb 2026 10:28:56 +0100
Subject: [PATCH] runtime: skip CDI guest annotations for non-VFIO devices

The siblingAnnotation function expects devPath to refer to a VFIO
device. If that's not the case, the function will fail to find the
sibling counterpart and return an error. This can be triggered by adding
a block device to the container.

The reason why this did not cause immediate problems is that the
error return was ignored on both call sites of
annotateContainerWithVFIOMetadata. This commit propagates the error
correctly.

Signed-off-by: Markus Rudy <mr@edgeless.systems>
---
 src/runtime/virtcontainers/container.go | 16 ++++++++++++++--
 1 file changed, 14 insertions(+), 2 deletions(-)

diff --git a/src/runtime/virtcontainers/container.go b/src/runtime/virtcontainers/container.go
index 1424a06ffab017be32369f649d52a739568cc515..032a97511c00c93301451ef180072ce0b2a87fe0 100644
--- a/src/runtime/virtcontainers/container.go
+++ b/src/runtime/virtcontainers/container.go
@@ -1004,7 +1004,9 @@ func (c *Container) createDevices(ctx context.Context, contConfig *ContainerConf
 
 	// If we're hot-plugging this will be a no-op because at this stage
 	// no devices are attached to the root-port or switch-port
-	c.annotateContainerWithVFIOMetadata(vfioColdPlugDevices)
+	if err := c.annotateContainerWithVFIOMetadata(vfioColdPlugDevices); err != nil {
+		return fmt.Errorf("annotating VFIO devices: %w", err)
+	}
 
 	return nil
 }
@@ -1117,6 +1119,10 @@ func (c *Container) annotateContainerWithVFIOMetadata(devices interface{}) error
 		// to the correct index
 		if devices, ok := devices.([]ContainerDevice); ok {
 			for _, dev := range devices {
+				if !strings.HasPrefix(dev.ContainerPath, "/dev/vfio") {
+					c.Logger().Infof("skipping guest annotations for non-VFIO device %q", dev.ContainerPath)
+					continue
+				}
 				if dev.ContainerPath == "/dev/vfio/vfio" {
 					c.Logger().Infof("skipping /dev/vfio/vfio for vfio_mode=guest-kernel")
 					continue
@@ -1138,6 +1144,10 @@ func (c *Container) annotateContainerWithVFIOMetadata(devices interface{}) error
 
 		if devices, ok := devices.([]config.DeviceInfo); ok {
 			for _, dev := range devices {
+				if !strings.HasPrefix(dev.ContainerPath, "/dev/vfio") {
+					c.Logger().Infof("skipping guest annotations for non-VFIO device %q", dev.ContainerPath)
+					continue
+				}
 				if dev.ContainerPath == "/dev/vfio/vfio" {
 					c.Logger().Infof("skipping /dev/vfio/vfio for vfio_mode=guest-kernel")
 					continue
@@ -1266,7 +1276,9 @@ func (c *Container) create(ctx context.Context) (err error) {
 		return
 	}
 
-	c.annotateContainerWithVFIOMetadata(c.devices)
+	if err = c.annotateContainerWithVFIOMetadata(c.devices); err != nil {
+		return
+	}
 
 	// Deduce additional system mount info that should be handled by the agent
 	// inside the VM
